internal/test/goleak/goleak_test.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) MongoDB, Inc. 2024-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.uber.org/goleak"
+)
+
+var dbName = fmt.Sprintf("goleak-%d", time.Now().Unix())
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
+}
+
+// TestGoroutineLeak creates clients with various client configurations, runs
+// some operations with each one, then disconnects the client. It asserts that
+// no goroutines were leaked after the client is disconnected.
+func TestGoroutineLeak(t *testing.T) {
+	uri := "mongodb://localhost:27017"
+	if u := os.Getenv("MONGODB_URI"); u != "" {
+		uri = u
+	}
+
+	testCases := []struct {
+		desc string
+		opts *options.ClientOptions
+	}{
+		{
+			desc: "base",
+			opts: options.Client().ApplyURI(uri),
+		},
+		{
+			desc: "compressors=snappy",
+			opts: options.Client().ApplyURI(uri).SetCompressors([]string{"snappy"}),
+		},
+		{
+			desc: "compressors=zlib",
+			opts: options.Client().ApplyURI(uri).SetCompressors([]string{"zlib"}),
+		},
+		{
+			desc: "compressors=zstd",
+			opts: options.Client().ApplyURI(uri).SetCompressors([]string{"zstd"}),
+		},
+		{
+			desc: "minPoolSize=10",
+			opts: options.Client().ApplyURI(uri).SetMinPoolSize(10),
+		},
+		{
+			desc: "serverMonitoringMode=poll",
+			opts: options.Client().ApplyURI(uri).SetServerMonitoringMode(options.ServerMonitoringModePoll),
+		},
+	}
+	for _, tc := range testCases {
+		// These can't be run in parallel because goleak currently can't filter
+		// out goroutines from other parallel subtests.
+		t.Run(tc.desc, func(t *testing.T) {
+			defer goleak.VerifyNone(t)
+
+			client, err := mongo.Connect(context.Background(), tc.opts)
+			require.NoError(t, err)
+
+			defer func() {
+				err = client.Disconnect(context.Background())
+				require.NoError(t, err)
+			}()
+
+			coll := client.Database(dbName).Collection(collectionName(t))
+			defer func() {
+				err := coll.Drop(context.Background())
+				require.NoError(t, err)
+			}()
+
+			_, err = coll.InsertOne(context.Background(), bson.M{"x": 123})
+			require.NoError(t, err)
+
+			for i := 0; i < 20; i++ {
```

Comment:
What is the significance of 20? Should we constantize this value to give it context?

Reply:
20 is "enough to cause a noticeable problem", so it doesn't have any special significance. However, note that we've only detected one type of goroutine leak with this test so far, so it's possible that we'll need to change the test scenario, duration, etc to catch future goroutine leaks. I'll leave a comment about the significance and limitations of the overall test loop related to detecting goroutine leaks. Ideally, we should actually be testing for goroutine leaks in our existing test suites, but that has proved somewhat difficult to implement.

Reply:
I ended up removing the sleep, increased the iterations to 50, moved the `InsertOne` into the loop, added a `ChangeStream` simulated workload, and added some operations that intentionally time out. I added a comment describing the intent of the different operations.